stan/math/prim/mat/prob/ordered_logistic_glm_lpmf.hpp

```diff
@@ -0,0 +1,174 @@
+
+  check_consistent_size(function, "Vector of dependent variables", y,
+                        N_instances);
+  check_consistent_size(function, "Weight vector", beta, N_attributes);
+  check_bounded(function, "Vector of dependent variables", y, 1, N_classes);
+  check_ordered(function, "Cut-points", cuts);
+  check_finite(function, "Final cut-point", cuts[N_classes - 2]);
+  check_finite(function, "First cut-point", cuts[0]);
+
+  if (size_zero(y, x, beta, cuts))
+    return 0;
+
+  if (!include_summand<propto, T_x_scalar, T_beta, T_cuts>::value)
+    return 0;
+
+  T_partials_return logp(0);
+  const auto& x_val = value_of_rec(x);
+  const auto& beta_val = value_of_rec(beta);
+  const auto& cuts_val = value_of_rec(cuts);
+
+  const auto& y_vec = as_column_vector_or_scalar(y);
+  const auto& beta_val_vec = as_column_vector_or_scalar(beta_val);
+  const auto& cuts_val_vec = as_column_vector_or_scalar(cuts_val);
+
+  Array<double, Dynamic, 1> cuts_y1(N_instances), cuts_y2(N_instances);
+  for (int i = 0; i < N_instances; i++) {
+    int c = y_vec[i];
+    if (c != N_classes) {
+      cuts_y1[i] = cuts_val_vec[c - 1];
+    } else {
+      cuts_y1[i] = INFINITY;
+    }
+    if (c != 1) {
+      cuts_y2[i] = cuts_val_vec[c - 2];
+    } else {
+      cuts_y2[i] = -INFINITY;
+    }
+  }
+
+  Array<double, Dynamic, 1> location = x_val * beta_val_vec;
+  if (!location.allFinite()) {
+    check_finite(function, "Weight vector", beta);
+    check_finite(function, "Matrix of independent variables", x);
+  }
+
+  Array<double, Dynamic, 1> cut2 = location - cuts_y2;
+  Array<double, Dynamic, 1> cut1 = location - cuts_y1;
+
+  // Not immediately evaluating next two expressions benefits performance
+  auto m_log_1p_exp_cut1
+      = -cut1 * (cut1 > 0.0).cast<double>() - (-cut1.abs()).exp().log1p();
+  auto m_log_1p_exp_m_cut2
+      = cut2 * (cut2 <= 0.0).cast<double>() - (-cut2.abs()).exp().log1p();
+  logp = y_vec.cwiseEqual(1)
+             .select(m_log_1p_exp_cut1,
+                     y_vec.cwiseEqual(N_classes).select(
+                         m_log_1p_exp_m_cut2,
+                         m_log_1p_exp_m_cut2 + log1m_exp(cut1 - cut2).array()
+                             + m_log_1p_exp_cut1))
+             .sum();
+
+  operands_and_partials<Matrix<T_x_scalar, Dynamic, Dynamic>, T_beta, T_cuts>
+      ops_partials(x, beta, cuts);
+  if (!is_constant_struct<T_x_scalar>::value
+      || !is_constant_struct<T_beta>::value
+      || !is_constant_struct<T_cuts>::value) {
+    Array<double, Dynamic, 1> d1
+        = 1 / (1 + exp(cut2)) - 1 / (1 - exp(cuts_y1 - cuts_y2));
+    Array<double, Dynamic, 1> d2
+        = 1 / (1 - exp(cuts_y2 - cuts_y1)) - 1 / (1 + exp(cut1));
+
+    if (!is_constant_struct<T_x_scalar>::value
+        || !is_constant_struct<T_beta>::value) {
+      Matrix<double, 1, Dynamic> location_derivative = d1 - d2;
+      if (!is_constant_struct<T_x_scalar>::value) {
+        ops_partials.edge1_.partials_
```

Comment:
Just checking that these shouldn't be `+=`.  We've run into that bug in other uses of `operands_and_partials` recently.

Reply:
No, `=` is correct here. This is not even in a loop, so there is no danger of overwriting previously calculated value.

Reply:
OK, thanks for checking.